File: zh-Hant/codes/cpp/chapter_hashing/hash_map.cpp
/**
 * File: hash_map.cpp
 * Created Time: 2022-12-14
 * Author: msk397 (machangxinq@gmail.com)
 */

#include "../utils/common.hpp"

/* Driver Code */
int main() {
    /* 初始化哈希表 */
    unordered_map<int, string> map;

    /* 添加操作 */
    // 在哈希表中添加键值对 (key, value)
    map[12836] = "小哈";
    map[15937] = "小啰";
    map[16750] = "小算";
    map[13276] = "小法";
    map[10583] = "小鸭";
    cout << "\n添加完成后，哈希表为\nKey -> Value" << endl;
    printHashMap(map);

    /* 查询操作 */
    // 向哈希表中输入键 key ，得到值 value
    string name = map[15937];
    cout << "\n输入学号 15937 ，查询到姓名 " << name << endl;

    /* 删除操作 */
    // 在哈希表中删除键值对 (key, value)
    map.erase(10583);
    cout << "\n删除 10583 后，哈希表为\nKey -> Value" << endl;
    printHashMap(map);

    /* 遍历哈希表 */
    cout << "\n遍历键值对 Key->Value" << endl;
    for (auto kv : map) {
        cout << kv.first << " -> " << kv.second << endl;
    }
    cout << "\n使用迭代器遍历 Key->Value" << endl;
    for (auto iter = map.begin(); iter != map.end(); iter++) {
        cout << iter->first << "->" << iter->second << endl;
    }

    return 0;
}

File: zh-hant/codes/dart/chapter_hashing/hash_map.dart
/**
 * File: hash_map.dart
 * Created Time: 2023-03-29
 * Author: liuyuxin (gvenusleo@gmail.com)
 */

/* Driver Code */
void main() {
  /* 初始化雜湊表 */
  final Map<int, String> map = {};

  /* 新增操作 */
  // 在雜湊表中新增鍵值對 (key, value)
  map[12836] = "小哈";
  map[15937] = "小囉";
  map[16750] = "小算";
  map[13276] = "小法";
  map[10583] = "小鴨";
  print("\n新增完成後，雜湊表為\nKey -> Value");
  map.forEach((key, value) => print("$key -> $value"));

  /* 查詢操作 */
  // 向雜湊表中輸入鍵 key ，得到值 value
  final String? name = map[15937];
  print("\n輸入學號 15937 ，查詢到姓名 $name");

  /* 刪除操作 */
  // 在雜湊表中刪除鍵值對 (key, value)
  map.remove(10583);
  print("\n刪除 10583 後，雜湊表為\nKey -> Value");
  map.forEach((key, value) => print("$key -> $value"));

  /* 走訪雜湊表 */
  print("\n走訪鍵值對 Key->Value");
  map.forEach((key, value) => print("$key -> $value"));
  print("\n單獨走訪鍵 Key");
  map.keys.forEach((key) => print(key));
  print("\n單獨走訪值 Value");
  map.forEach((key, value) => print("$value"));
  map.values.forEach((value) => print(value));
}
